ompi/mca/pml/monitoring/pml_monitoring.c

```diff
@@ -0,0 +1,228 @@
+/*
+ * Copyright (c) 2013-2015 The University of Tennessee and The University
+ *                         of Tennessee Research Foundation.  All rights
+ *                         reserved.
+ * Copyright (c) 2013-2015 Inria.  All rights reserved.
+ * Copyright (c) 2015      Bull SAS.  All rights reserved.
+ * $COPYRIGHT$
+ *
+ * Additional copyrights may follow
+ *
+ * $HEADER$
+ */
+
+#include <ompi_config.h>
+#include <pml_monitoring.h>
+#include "opal/class/opal_hash_table.h"
+
+int filter_monitoring( void );
+
+/* array for stroring monitoring data*/
+uint64_t* sent_data = NULL;
+uint64_t* messages_count = NULL;
+uint64_t* filtered_sent_data = NULL;
+uint64_t* filtered_messages_count = NULL;
+
+static int init_done = 0;
+static int nbprocs = -1;
+static int my_rank = -1;
+opal_hash_table_t *translation_ht = NULL;
+
+
+mca_pml_monitoring_module_t mca_pml_monitoring = {
+    mca_pml_monitoring_add_procs,
+    mca_pml_monitoring_del_procs,
+    mca_pml_monitoring_enable,
+    NULL,
+    mca_pml_monitoring_add_comm,
+    mca_pml_monitoring_del_comm,
+    mca_pml_monitoring_irecv_init,
+    mca_pml_monitoring_irecv,
+    mca_pml_monitoring_recv,
+    mca_pml_monitoring_isend_init,
+    mca_pml_monitoring_isend,
+    mca_pml_monitoring_send,
+    mca_pml_monitoring_iprobe,
+    mca_pml_monitoring_probe,
+    mca_pml_monitoring_start,
+    mca_pml_monitoring_improbe,
+    mca_pml_monitoring_mprobe,
+    mca_pml_monitoring_imrecv,
+    mca_pml_monitoring_mrecv,
+    mca_pml_monitoring_dump,
+    NULL,
+    65535,
+    INT_MAX
+};
+
+int mca_pml_monitoring_add_procs(struct ompi_proc_t **procs,
+                                 size_t nprocs)
+{
+    /**
+     * Create the monitoring hashtable only for my MPI_COMM_WORLD. We choose
+     * to ignore by now all other processes.
+     */
+    if(NULL == translation_ht) {
+        size_t i;
+        uint64_t key;
+
+        nbprocs = nprocs;
+
+        translation_ht = OBJ_NEW(opal_hash_table_t);
+        opal_hash_table_init(translation_ht, 2048);
+
+
+        for( i = 0; i < nprocs; i++ ) {
+            /* rank : ompi_proc_local_proc in procs */
+            if( procs[i] == ompi_proc_local_proc)
+                my_rank = i;
+            key = *((uint64_t*)&(procs[i]->super.proc_name));
+            /* store the rank (in COMM_WORLD) of the process
+               with its name (a uniq opal ID) as key  in the hash table*/
+            opal_hash_table_set_value_uint64(translation_ht,
+                                             key,
+                                             (void*)(uintptr_t)i);
+        }
+    }
+    return pml_selected_module.pml_add_procs(procs, nprocs);
```

Comment:
Keep in mind that when dynamic add_procs is in use the above code will only add the local ranks to translation_ht. It looks like you don't need the procs, just the opal_process_name_t's. You can generate those yourself as is done at ompi/proc/proc.c:337 .